### ISSUE TITLE
merTools is unaware of user specified formulas

### ISSUE THREAD

**Issue by jknowles** [jknowles/merTools#59]
Thanks to @cchoirat for reporting this:

``` r
set.seed(271828)
data(sleepstudy)
```

``` r
fm1 <- lmer(Reaction ~ Days + (Days|Subject), data=sleepstudy)
draw(fm1, type = 'average')
```

```
Number of observations < 20, random effect quantiles may not be well-defined.
  Reaction Days Subject
1 298.5079  4.5     308
```

``` r
fm2 <- lmer(Reaction ~ Days + I(Days^2) + (Days|Subject), data=sleepstudy)
draw(fm2, type = 'average')
```

```
Number of observations < 20, random effect quantiles may not be well-defined.
  Reaction Days I(Days^2) Subject
1 298.5079  4.5      28.5     308
```

I(Days^2) is 28.5.  It should be 20.25.

Right now `draw` is unaware of any formulas and is independently computing the mean of the `Days` and the `I(Days^2)` vectors, which is causing this issue.

**Comment by carlbfrederick** [comment]
As I have looked into this issue, this brought three broad scenarios to mind: 

  - formulas with `I()` variable constructions
  - formulas with other functions (e.g. `log`, `sqrt`) variable constructions
  - formulas with interactions (`:` or `*`)

Turns out the interactions are fine, the first two scenarios are potentially problematic as noted in the original issue and potentially even more so.

**IF** the 'main effects' variable is present in the model frame, we could write a second function that could call `collapseFrame()` and make the appropriate transformations.

**Without the 'main effects'**, I am not sure we could recover data from the model.frame.  

Options on how to proceed:

  1.  Issue a warning and put in doco when functions are found in the formula that user should check to make sure they get the correct average and edit if not.
  2. Modify `shinyMer` to accept edits (via DT) to when they select averageObs
  3. We could automate such calculations when 'main effects' are present for folks

thoughts?

**Comment by jknowles** [comment]
That sounds sensible to me. 

Formula parsing is a nightmare. We'll also want to test we get it right in these cases too:

```
y ~ x1 + x2 + (1 + sqrt(x1)|g1)
y ~ x1 + factor(x2) + (1 + factor(x2)|g1)
y ~ x1 + I(x1^2) + (1  + x1|g1)
```

Just to cover our bases because users may put transformations in the random effects groups as well. 

What do you think?

**Comment by carlbfrederick** [comment]
True, who knows where these goofy things might end up :)  Luckily, it is just as easy to extract the variable names from the random part of the merMod as it is from the fixed part.  I have been using `all.vars` to get an unduplicated vector of varnames and then we can just `do.call` the functions by extracting the names of the model.frame or the from the `terms()` object.

factor would only become a problem in data that doesn't come from the model.frame, but that would be easy to scan for and adjust post-hoc as well, again as long as the original variable is there.

I am leaning towards option 1 definitely and option 2 if it isn't too hard ... could also be it's own separate option for shinyMer that is pre-populated with averageObs.  I think we can just scrap option 3 and add advice about creating normal variables in the first place if they really require such oddball formula specifications and edit any dependencies themselves.

**Comment by jknowles** [comment]
That seems like the best approach to me too. It's mostly important to catch it so users don't mistakenly think they are getting valid predictions for squared terms and the like if they input them into the formula interface.

**Comment by carlbfrederick** [comment]
@jknowles check out SHA: d18d3667f2ab1340637295c07576c3690a69cedc on the issue59 branch for the direction I am going for the fix of #59  

I should be able to finish up tomorrow.
